gh-136459: Add perf trampoline support for macOS

canova opened the pull request:
This PR adds perf trampoline support for macOS. Previously it was Linux-only, which is not actually needed it to be. We can share the implementation of Linux and macOS with some minor changes done in this PR.

For example, here's a before and after profile of this PR, captured by `samply` (for the same dummy script):
[Before](https://share.firefox.dev/3GCwhvp) / [After](https://share.firefox.dev/461yivE)

As you can see, now we have blue frames that mentions the real Python function names, which makes it a lot more useful to understand this profile. Without it, it's just native frames that's really difficult to understand.

This is my first PR in this project, so please let me know if there is anything I didn't do or should do. Thanks!

<!-- gh-issue-number: gh-136459 -->
* Issue: gh-136459
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--136461.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

canova committed:
Add perf trampoline support for macOS

canova committed:
Make sure that test_perfmaps.py test is not skipped on macOS

canova committed:
Update the docs for perfmaps to mention that macOS is supported

canova committed:
Add myself to Misc/ACKS

canova committed:
Add a Misc/NEWS.d entry

canova commented on Doc/c-api/perfmaps.rst:
```diff
@@ -5,11 +5,12 @@
-A running process may create a file in the ``/tmp`` directory, which contains entries
-that can map a section of executable code to a name. This interface is described in the
+profiling tool (such as `perf <https://perf.wiki.kernel.org/index.php/Main_Page>`_ or
+`samply <https://github.com/mstange/samply/>`_). A running process may create a
```
I added a shameless plug to samply 😄 Disclaimer: I'm not the maintainer of the project, but the maintainer is my colleague. But it doesn't change the fact that it's an awesome profiler! But I can revert it if you prefer not to include :)

pablogsal replied:
I am happy with the plug, but this docs are going to need much more than this then. If `samply` is the main way to use this on macOS then we will need to update https://docs.python.org/3/howto/perf_profiling.html with full instructions for samply :)

canova commented on configure.ac:
```diff
@@ -3690,12 +3690,13 @@ case "$ac_sys_system" in
 AS_CASE([$PLATFORM_TRIPLET],
   [x86_64-linux-gnu], [perf_trampoline=yes],
   [aarch64-linux-gnu], [perf_trampoline=yes],
+  [darwin], [perf_trampoline=yes],
```
Apparently `PLATFORM_TRIPLET` is not a triplet on macOS. But this is fine for this case because both x86_64 and aarch64 is supported.

Zheaoli commented on Python/perf_jit_trampoline.c:
```diff
@@ -102,6 +106,16 @@
+
+/* These constants are defined inside <elf.h>, which we can't use outside of linux. */
+#if !defined(__linux__)
+#  define EM_386      3
```
I think we need to define the variable for different platforms. like this https://github.com/python/cpython/blob/main/Python/perf_jit_trampoline.c#L126-L135

Not define all the possible variable

canova replied:
Sure, I just changed them to be defined depending on the platform. I actually copy pasted this from the old code but your suggestion is better.

Zheaoli reviewed (commented):
For now, the trampoline depens on the mmap.

But it has record about the mmap performance issue on macOS. 

I think we need a benchmark here

FYI 
1. https://github.com/tmm1/node/commit/3542e4e2944f2fe1132ded1052e223f54c90e4bf
2. https://chromium-review.googlesource.com/c/v8/v8/+/6385655/7/src/diagnostics/perf-jit.cc
3. https://bugzilla-dev.allizom.org/show_bug.cgi?id=1827214

canova committed:
Define constants per-platform

canova committed:
Do not mmap the jitdump file on macOS

On macOS, we don't need to call mmap because samply has already detected
the file path during the call to `open` before (it interposes `open` with
a preloaded library), and because the mmap call can be slow.

pablogsal commented on configure.ac:
```diff
@@ -3690,12 +3690,13 @@ case "$ac_sys_system" in
 AS_CASE([$PLATFORM_TRIPLET],
   [x86_64-linux-gnu], [perf_trampoline=yes],
   [aarch64-linux-gnu], [perf_trampoline=yes],
+  [darwin], [perf_trampoline=yes],
```
Will this trigger also on IOS @freakboy3742 ? We only want macOS here

freakboy3742 replied:
No - on iOS, `PLATFORM_TRIPLET` will be `arm64-iphoneos`, `arm64-iphonesimulator` or `x86_64-iphonesimulator`, so this won't match.

pablogsal commented on Python/asm_trampoline.S:
```diff
@@ -1,13 +1,21 @@
     .text
+#if defined(__APPLE__)
+    .globl	__Py_trampoline_func_start
```
Ah, Apple and their symbol choices...:S

corona10 commented on Python/perf_jit_trampoline.c:
```diff
@@ -1166,7 +1191,11 @@ static void perf_map_jit_write_entry(void *state, const void *code_addr,
     ev.base.time_stamp = get_current_monotonic_ticks();
     ev.process_id = getpid();
+#if defined(__APPLE__)
+    pthread_threadid_np(NULL, &ev.thread_id);
```
Can we just use this implementation in here?
https://github.com/python/cpython/blob/b44316a0976fb3fcd50bae9d67b0810ee0252d93/Include/object.h#L193-L255

canova replied:
Thanks for the suggestion! It looks like `_Py_ThreadId` is defined only when `Py_GIL_DISABLED` is defined and `Py_LIMITED_API` is not defined:
https://github.com/python/cpython/blob/d754f75f42f040267d818ab804ada340f55e5925/Include/object.h#L189
It doesn't compile because of that.

corona10 replied:
Ah I meant just for inline assembly because it was for low over head at free threading, maybe you can just pick it for apple implementation.

canova committed:
Update the perf profiling doc to include samply

canova committed:
Add some tests for samply profiling

pablogsal reviewed (commented):
@hugovk @AA-Turner I would love if you could help us a bit with the docs. The previous docs are assuming that only the perf profiler exist and is centered on it but seems we want to move to a world when perf and other perf-like profilers can exist. I want this to be a excellent place to learn about this so I would love if you could guide @canova on how to ensure there is enough advice on how to use "simply" that users can have a good experience and doesn't feel out of place.

pablogsal commented on Doc/howto/perf_profiling.rst:
```diff
@@ -148,6 +149,26 @@ Instead, if we run the same experiment with ``perf`` support enabled we get:
 
 
 
+Using ``samply`` profiler
```
We are going to need a bit more here. For example, simply supports both perf modes so we need clarification on when tho use them and what are the recommendations. How to read the flamegraphs etc etc

mstange replied:
Would it make sense to break these discussions out into a separate PR? It doesn't seem useful to delay landing trampoline support for this.

pablogsal replied:
> It doesn't seem useful to delay landing trampoline support for this.

Is there any rush? This will go into 3.15 anyway and that's going to be released October 2026. We still need to figure out the buildbot situation which will take some time...

pablogsal replied:
I am happy to separate this into a different PR, though

canova replied:
Oh, I was hoping that we could maybe enable it on 3.14. Considering that the code was there since 3.12, and it's mostly putting lots of ifdefs here and there (minus samply and documentation part). I suspect that updating the documentation will take longer. But I'm not familiar with the release process.

pablogsal replied:
> Oh, I was hoping that we could maybe enable it on 3.14.

No way unfortunately  as we are 3 betas past beta freeze. It's up to the release manager to decide (CC @hugovk) but we have a strict policy for this I am afraid and no new features can be added past beta freeze.

pablogsal replied:
@hugovk Checking just in case although I assume the answer is "no" but would you consider adding this to 3.14 given that this is a new platform and the code is gated by ifdefs? This will allow people on macOS to profile their code using a native profiler, which would be very useful for investigating performance in Python+compiled code.

pablogsal replied:
Some context for this: this would allow people on macOS to profile free threaded Python using samply, so maybe there is a case to allow it in 3.14 but I am still unsure. Up to you @hugovk

hugovk replied:
I think we're a bit too close to RC to make an exception for this.

canova replied:
Aww, I'm a bit sad to see that I have to wait for more than 1 year to be able to profile performance of python scripts on macOS. I know some people were excited to use it on macOS at Mozilla to improve the performance of our build system. But thanks for taking time and answering!

I could argue that this PR is extending support of an existing feature, and not exactly adding a new feature 😄  And this code is not executed as long as `PYTHONPERFSUPPORT=1` or `-X perf` is not passed, so I think it's low risk. But I understand that it's close to RC too.

(thanks for the doc suggestions, I'll update the PR soon!)

hugovk commented on Doc/c-api/perfmaps.rst:
```diff
@@ -5,11 +5,12 @@
 ----------------------
 
-On supported platforms (as of this writing, only Linux), the runtime can take
+On supported platforms (as of this writing, Linux and macOS), the runtime can take
```
Everything in this versioned documentation is as of this writing :)
```suggestion
On supported platforms (Linux and macOS), the runtime can take
```

hugovk commented on Doc/howto/perf_profiling.rst:
```diff
@@ -2,34 +2,35 @@
-    select architectures. Check the output of the ``configure`` build step or
+    Support for profiling is available on Linux and macOS on select architectures.
+    ``perf`` is available on Linux, while ``samply`` can be used on both Linux and macOS.
+    ``samply`` support on macOS is available starting from Python 3.14.
```
```suggestion
    Perf is available on Linux, while samply can be used on both Linux and macOS.
    samply support on macOS is available starting from Python 3.14.
```

hugovk commented on Doc/howto/perf_profiling.rst:
```diff
@@ -148,6 +149,26 @@ Instead, if we run the same experiment with ``perf`` support enabled we get:
 
 
+Using ``samply`` profiler
+-------------------------
```
```suggestion
Using the samply profiler
-------------------------
```

hugovk commented on Doc/howto/perf_profiling.rst:
```diff
@@ -148,6 +149,26 @@ Instead, if we run the same experiment with ``perf`` support enabled we get:
+``samply`` is a modern profiler that can be used as an alternative to ``perf``.
+It uses the same perf map files that Python generates, making it compatible
+with Python's profiling support. ``samply`` is particularly useful on macOS
+where ``perf`` is not available.
```
```suggestion
samply is a modern profiler that can be used as an alternative to perf.
It uses the same perf map files that Python generates, making it compatible
with Python's profiling support. samply is particularly useful on macOS
where perf is not available.
```

hugovk commented on Doc/howto/perf_profiling.rst:
```diff
@@ -148,6 +149,26 @@ Instead, if we run the same experiment with ``perf`` support enabled we get:
+with Python's profiling support. ``samply`` is particularly useful on macOS
+where ``perf`` is not available.
+
+To use ``samply`` with Python, first install it following the instructions at
```
```suggestion
To use samply with Python, first install it following the instructions at
```

hugovk commented on Doc/howto/perf_profiling.rst:
```diff
@@ -148,6 +149,26 @@ Instead, if we run the same experiment with ``perf`` support enabled we get:
+web-based interface for analyzing profiling data and works on both Linux
+and macOS.
+
+On macOS, ``samply`` support requires Python 3.14 or later.
```
```suggestion
interactively. The advantage of samply is that it provides a modern
web-based interface for analyzing profiling data and works on both Linux
and macOS.

On macOS, samply support requires Python 3.14 or later.
```

hugovk commented on Misc/NEWS.d/next/Core_and_Builtins/2025-07-09-11-15-42.gh-issue-136459.m4Udh8.rst:
```diff
@@ -0,0 +1,3 @@
+Add support for perf trampoline on macOS, to allow profilers wit JIT map
+support to read python calls. While profiling, ``PYTHONPERFSUPPORT=1`` can
```
```suggestion
support to read Python calls. While profiling, ``PYTHONPERFSUPPORT=1`` can
```

hugovk reviewed (commented):
> @hugovk @AA-Turner I would love if you could help us a bit with the docs. The previous docs are assuming that only the perf profiler exist and is centered on it but seems we want to move to a world when perf and other perf-like profilers can exist. I want this to be a excellent place to learn about this so I would love if you could guide @canova on how to ensure there is enough advice on how to use "simply" that users can have a good experience and doesn't feel out of place.

Good idea, here's a few brief docs comments to begin with, we can perhaps do a rewrite in a followup?

canova committed:
Apply documentation suggestions from code review

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

canova committed:
Change the version number in the docs and mention macOS restrictions
